common/predictive-text/virtualizedWorker.ts

```diff
@@ -0,0 +1,86 @@
+/// <reference path="embedded_worker.d.ts" />
+/// <reference path="worker-interface.d.ts" />
+
+namespace com.keyman.text.prediction {
+  // Unfortunately, importing like this loses all type information for the modules.
+  // It's that, or massively overhaul lm-layer to be module-based instead of
+  // namespace-based.
+  var fs = require("fs");
+  var vm = require("vm");
+
+  class VirtualizedWorkerContext {
+    // The LMLayerWorker installs itself to 'self', the expected Worker global, so we provide an alias.
+    self: VirtualizedWorkerContext;
+
+    constructor() {
+      this.self = this;
+    }
+
+    postMessage: (message: any, options: PostMessageOptions) => void;
+
+    importScripts() {
+      /* Use of vm.createContext and script.runInContext allow us to avoid
+        * polluting the global scope with imports.  When we throw away the
+        * context object, imported scripts will be automatically GC'd.
+        */
+      for(var i=0; i < arguments.length; i++) {
+        this.__importScriptString(fs.readFileSync(arguments[i]) as any as string);
+      }
+    }
+
+    __importScriptString(scriptStr: string) {
+      let context = vm.createContext(this);
+      var script = new vm.Script(scriptStr);
+      script.runInContext(context);
+    }
+  }
+
+  /**
+   * Note:  this does not create an actual Worker, separate process, or thread.  Everything will
+   *        be executed in-line on a virtualized context.
+   * 
+   *        In the future, it might be nice to use Node's Worker Threads implementation.
+   */
+  export class VirtualizedWorker implements Worker {
+    private _workerContext: VirtualizedWorkerContext;
+
+    constructor(scriptStr: string) {
+      this._workerContext = new VirtualizedWorkerContext();
+      // Needs to exist before setting up the worker; must exist by `.install()`.
+      this._workerContext.postMessage = this.workerPostMessage.bind(this);
+
+      // Initialize the "worker".
+      this._workerContext.__importScriptString(scriptStr);
+    }
+
+    // Sends the worker's postMessage messages to the appropriate `onmessage` handler.
+    private workerPostMessage(message: any, options: any) {
+      if(this.onmessage) {
+        this.onmessage({data: message} as any as MessageEvent);
+      }
+    }
+
+    onmessage: (this: Worker, ev: MessageEvent) => any;
```

Comment:
I forgot why this class does not implement `onmessage()`, but then I remembered how my own code works! :joy:

I suggest having a comment similar to this one to avoid future confusion:

```suggestion
    /* This is assigned by LMLayerWorker.install() during initialization */
    onmessage: (this: Worker, ev: MessageEvent) => any;
```